/*
 *  Copyright (c) 2025, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

syntax = "proto3";

package wso2.gateway.python.v1;

option go_package = "github.com/wso2/api-platform/gateway/gateway-runtime/policy-engine/internal/pythonbridge/proto";

import "google/protobuf/struct.proto";

// PythonExecutorService defines the gRPC contract between Go PE and the Python process.
// The Python process is the gRPC SERVER, Go PE is the CLIENT.
service PythonExecutorService {
  // Bidirectional stream for executing policies.
  // Go sends ExecutionRequest, Python responds with ExecutionResponse.
  // Each request has a unique request_id that the response must echo back.
  rpc ExecuteStream (stream ExecutionRequest) returns (stream ExecutionResponse);

  // Health check for readiness.
  rpc HealthCheck (HealthCheckRequest) returns (HealthCheckResponse);
}

// ---------------------- Request / Response ----------------------

message ExecutionRequest {
  // Unique ID per call so responses can be correlated on the single stream.
  string request_id = 1;

  // Policy to execute (name:version format used for cache lookup, e.g., "my-policy:v1")
  string policy_name = 2;
  string policy_version = 3;

  // Phase: "on_request" or "on_response"
  string phase = 4;

  // Merged parameters (system + user) for this policy instance.
  // The Go side resolves ${config} references in systemParameters and merges
  // them with user parameters before sending. Python never sees raw ${config} strings.
  google.protobuf.Struct params = 5;

  // The request or response context data
  oneof context {
    RequestContext request_context = 6;
    ResponseContext response_context = 7;
  }

  // Shared context (metadata, API info, auth context)
  SharedContext shared_context = 8;

  // Policy metadata (route info, API info) for factory creation
  PolicyMetadata policy_metadata = 9;
}

message ExecutionResponse {
  // Must match request_id from the corresponding ExecutionRequest
  string request_id = 1;

  oneof result {
    RequestActionResult request_result = 2;
    ResponseActionResult response_result = 3;
    ExecutionError error = 4;
  }

  // Updated shared metadata (Python may have mutated it).
  // Go side merges this back into the SharedContext.
  google.protobuf.Struct updated_metadata = 5;
}

// ---------------------- Context Messages ----------------------

message SharedContext {
  string project_id = 1;
  string request_id = 2;
  google.protobuf.Struct metadata = 3;     // Inter-policy communication map
  string api_id = 4;
  string api_name = 5;
  string api_version = 6;
  string api_kind = 7;
  string api_context = 8;
  string operation_path = 9;
  map<string, string> auth_context = 10;
}

message RequestContext {
  map<string, string> headers = 1;
  bytes body = 2;
  bool body_present = 3;
  bool end_of_stream = 4;
  string path = 5;
  string method = 6;
  string authority = 7;
  string scheme = 8;
}

message ResponseContext {
  // Original request data (immutable)
  map<string, string> request_headers = 1;
  bytes request_body = 2;
  string request_path = 3;
  string request_method = 4;

  // Response data
  map<string, string> response_headers = 5;
  bytes response_body = 6;
  bool response_body_present = 7;
  int32 response_status = 8;
}

message PolicyMetadata {
  string route_name = 1;
  string api_id = 2;
  string api_name = 3;
  string api_version = 4;
  string attached_to = 5;   // "api" or "route"
}

// ---------------------- Action Results ----------------------

message RequestActionResult {
  oneof action {
    UpstreamRequestModifications continue_request = 1;
    ImmediateResponseAction immediate_response = 2;
  }
}

message ResponseActionResult {
  oneof action {
    UpstreamResponseModifications continue_response = 1;
  }
}

message UpstreamRequestModifications {
  map<string, string> set_headers = 1;
  repeated string remove_headers = 2;
  map<string, StringList> append_headers = 3;
  bytes body = 4;
  bool body_present = 5;              // false means no body change, true means use body field (even if empty)
  string path = 6;
  bool path_present = 7;
  string method = 8;
  bool method_present = 9;
  google.protobuf.Struct analytics_metadata = 10;
}

message UpstreamResponseModifications {
  map<string, string> set_headers = 1;
  repeated string remove_headers = 2;
  map<string, StringList> append_headers = 3;
  bytes body = 4;
  bool body_present = 5;
  int32 status_code = 6;
  bool status_code_present = 7;
  google.protobuf.Struct analytics_metadata = 8;
}

message ImmediateResponseAction {
  int32 status_code = 1;
  map<string, string> headers = 2;
  bytes body = 3;
  google.protobuf.Struct analytics_metadata = 4;
}

message ExecutionError {
  string message = 1;
  string policy_name = 2;
  string policy_version = 3;
  string error_type = 4;    // "init_error", "execution_error", "timeout"
}

// ---------------------- Health Check ----------------------

message HealthCheckRequest {}

message HealthCheckResponse {
  bool ready = 1;
  int32 loaded_policies = 2;
}

// ---------------------- Utility ----------------------

message StringList {
  repeated string values = 1;
}
